src/drivers/intel/cavs/ipc.c

```diff
@@ -185,6 +194,15 @@ enum task_state ipc_platform_do_cmd(struct ipc *ipc)
 	hdr = mailbox_validate();
 #endif
 
+#ifdef IPC4_EARLY_ACK
+	/*
+	 * notify host that we have received the message and
+	 * do this before we send the response
+	 */
+	ipc_write(IPC_DIPCTDR, ipc_read(IPC_DIPCTDR) | IPC_DIPCTDR_BUSY);
+	ipc_write(IPC_DIPCTDA, ipc_read(IPC_DIPCTDA) | IPC_DIPCTDA_DONE);
```

Comment:
According to cAVS 2.5 HW Arch Spec, IPC flow should be implemented as follows:
1. DSP receives the interrupt, reads message from DIPCTDR, DIPCTDD and L2 local memory
2. DSP clears BUSY by writing 1 to DIPCTDR
3. DSP process the message
4. DSP set DONE by writing DIPCTDA

Replacing 3. and 4. will violate the architecture flow.
Note: in closed-source cAVS, BUSY is cleared together with set DONE, after processing the message, so moment of clearing BUSY seems to not to be critical.

At present cAVS closed-source solution, BUSY and DONE are triggered after processing message from host and before writing response. It seems to be the best way.

Reply:
Thanks @pblaszko ! Indeed, this is what @ujfalusi  also commented in last Friday in https://github.com/thesofproject/sof/pull/6239#issuecomment-1235090106 . And yes, this seems to be the correct thing.

The key seems to be whether busy+done clearance is done before or after the response is sent. We are err'ing towards the direction that host/kernel should handle even the case where busy+done is cleared only at the end (i.e. the original sequence).